libavfilter/vf_signature.c

```diff
@@ -748,6 +758,310 @@ static int config_output(AVFilterLink *outlink)
+	sc->time_base.num = 1;
+
+	// Skip the following data
+	// - MediaTimeFlagOfSpatialRegion: (1 bit) always 1
+	// - StartMediaTimeOfSpatialRegion: (32 bits) always 0
+	skip_bits(&bitContext, 1 + 32);
+
+	// EndMediaTimeOfSpatialRegion
+	uint64_t lastCoarsePts = get_bits_long(&bitContext, 32);
+
+	// Coarse signatures
+	// numOfSegments = number of coarse signatures
+	numOfSegments = get_bits_long(&bitContext, 32);
+
+	// Reading from binary signature return a wrong number of segments
+	// numOfSegments = (sc->lastindex + 44)/45;
+	//skip_bits(&bitContext, 32);
+
+	sc->coarsesiglist = (CoarseSignature*)calloc(numOfSegments,	sizeof(CoarseSignature));
+
+	BoundedCoarseSignature *bCoarseList = (BoundedCoarseSignature*)calloc(numOfSegments, sizeof(BoundedCoarseSignature));
+
+	// CoarseSignature loading
+	for (unsigned int i = 0; i < numOfSegments; ++i) {
+		BoundedCoarseSignature *bCs = &bCoarseList[i];
+		bCs->cSign = &sc->coarsesiglist[i];
+
+		if (i < numOfSegments - 1)
+			bCs->cSign->next = &sc->coarsesiglist[i + 1];
+		// each coarse signature is a VSVideoSegment
+		// StartFrameOfSegment
+		bCs->firstIndex = get_bits_long(&bitContext, 32);
+		// EndFrameOfSegment
+		bCs->lastIndex = get_bits_long(&bitContext, 32);
+
+		// MediaTimeFlagOfSegment 1 bit, always 1
+		skip_bits(&bitContext, 1);
+
+		// Fine signature pts
+		// StartMediaTimeOfSegment 32 bits
+		bCs->firstPts = get_bits_long(&bitContext, 32);
+		// EndMediaTimeOfSegment 32 bits
+		bCs->lastPts = get_bits_long(&bitContext, 32);
+		// Bag of words
+		for (unsigned int i = 0; i < 5; ++i) {
+
+			// read 243 bits ( = 7 * 32 + 19 = 8 * 28 + 19) into buffer
+			for (unsigned int j = 0; j < 30; ++j) {
+				// 30*8 bits = 30 bytes
+				bCs->cSign->data[i][j] = get_bits(&bitContext, 8);
+			}
+			bCs->cSign->data[i][30] = get_bits(&bitContext, 3) << 5;
+		}
+	}
+	sc->coarseend = &sc->coarsesiglist[numOfSegments - 1];
+
+	// Finesignatures
+	// CompressionFlag, only 0 supported
+	skip_bits(&bitContext, 1);
+
+	sc->finesiglist = (FineSignature*)calloc(sc->lastindex,	sizeof(FineSignature));
+
+	// Load fine signatures from file
+	for (unsigned int i = 0; i < sc->lastindex; ++i) {
+		FineSignature *fs = &sc->finesiglist[i];
+
+		// MediaTimeFlagOfFrame always 1
+		skip_bits(&bitContext, 1);
+
+		// MediaTimeOfFrame (PTS)
+		fs->pts = get_bits_long(&bitContext, 32);
+
+		// FrameConfidence
+		fs->confidence = get_bits(&bitContext, 8);
+
+		// words
+		for (unsigned int l = 0; l < 5; l++) {
+			fs->words[l] = get_bits(&bitContext, 8);
+		}
+
+		// Crashes for some signature, it's a memory adding problems
+		// framesignature
+		for (unsigned int l = 0; l < SIGELEM_SIZE / 5; l++) {
+			fs->framesig[l] = get_bits(&bitContext, 8);
+		}
+	}
+
+	// Sort by frame time (pts)
+	qsort(sc->finesiglist, sc->lastindex, sizeof(FineSignature),finesignaturecompare);
+
+	// Creating FineSignature linked list
+	for (unsigned int i = 0; i < sc->lastindex; ++i) {
+		FineSignature *fs = &sc->finesiglist[i];
```

Comment:
From what I understand during `binary_export` we traverse the linked list to write each fine signature to the file, as opposed to indexing over the array `s->finesiglist[i]`:
https://github.com/livepeer/FFmpeg/blob/b344789a31a2d806be702239d7614f7a9f51f941/libavfilter/vf_signature.c#L549

Shouldn't we read similarly - i.e. every new signature we read from the file goes into the `fs->next` of the previous signature? Why is the sorting done before building the LL?

Reply:
Good catch, Will fix!